Add default repr for EAs

TomAugspurger opened the pull request:
Closes https://github.com/pandas-dev/pandas/issues/22846
Closes https://github.com/pandas-dev/pandas/issues/23590

```python
In [4]: pd.core.arrays.period_array(['2000', '2001', None], freq='D')
Out[4]:
<PeriodArray>
['2000-01-01', '2001-01-01', 'NaT']
Length: 3, dtype: period[D]

In [5]: pd.core.arrays.period_array(['2000', '2001', None] * 100, freq='D')
Out[5]:
<PeriodArray>
['2000-01-01', '2001-01-01',        'NaT', '2000-01-01', '2001-01-01',
        'NaT', '2000-01-01', '2001-01-01',        'NaT', '2000-01-01',
 ...
        'NaT', '2000-01-01', '2001-01-01',        'NaT', '2000-01-01',
 '2001-01-01',        'NaT', '2000-01-01', '2001-01-01',        'NaT']
Length: 300, dtype: period[D]

In [6]: pd.core.arrays.integer_array([1, 2, None])
Out[6]:
<IntegerArray>
[1, 2, NaN]
Length: 3, dtype: Int64

In [7]: pd.core.arrays.integer_array([1, 2, None] * 1000)
Out[7]:
<IntegerArray>
[  1,   2, NaN,   1,   2, NaN,   1,   2, NaN,   1,
 ...
 NaN,   1,   2, NaN,   1,   2, NaN,   1,   2, NaN]
Length: 3000, dtype: Int64

```

TomAugspurger committed:
wip

pep8speaks commented:
Hello @TomAugspurger! Thanks for submitting the PR.

 - There are no PEP8 issues in the file [`pandas/core/arrays/base.py`](https://github.com/TomAugspurger/pandas/blob/0fdbfd34add4fe9940cd1991122e851d62824724/pandas/core/arrays/base.py) !

 - There are no PEP8 issues in the file [`pandas/core/arrays/categorical.py`](https://github.com/TomAugspurger/pandas/blob/0fdbfd34add4fe9940cd1991122e851d62824724/pandas/core/arrays/categorical.py) !

 - There are no PEP8 issues in the file [`pandas/core/arrays/integer.py`](https://github.com/TomAugspurger/pandas/blob/0fdbfd34add4fe9940cd1991122e851d62824724/pandas/core/arrays/integer.py) !

 - There are no PEP8 issues in the file [`pandas/core/arrays/interval.py`](https://github.com/TomAugspurger/pandas/blob/0fdbfd34add4fe9940cd1991122e851d62824724/pandas/core/arrays/interval.py) !

 - There are no PEP8 issues in the file [`pandas/core/arrays/period.py`](https://github.com/TomAugspurger/pandas/blob/0fdbfd34add4fe9940cd1991122e851d62824724/pandas/core/arrays/period.py) !

 - There are no PEP8 issues in the file [`pandas/io/formats/printing.py`](https://github.com/TomAugspurger/pandas/blob/0fdbfd34add4fe9940cd1991122e851d62824724/pandas/io/formats/printing.py) !

 - There are no PEP8 issues in the file [`pandas/tests/arrays/interval/test_interval.py`](https://github.com/TomAugspurger/pandas/blob/0fdbfd34add4fe9940cd1991122e851d62824724/pandas/tests/arrays/interval/test_interval.py) !

 - There are no PEP8 issues in the file [`pandas/tests/arrays/test_integer.py`](https://github.com/TomAugspurger/pandas/blob/0fdbfd34add4fe9940cd1991122e851d62824724/pandas/tests/arrays/test_integer.py) !

 - There are no PEP8 issues in the file [`pandas/tests/arrays/test_period.py`](https://github.com/TomAugspurger/pandas/blob/0fdbfd34add4fe9940cd1991122e851d62824724/pandas/tests/arrays/test_period.py) !

 - There are no PEP8 issues in the file [`pandas/tests/extension/base/__init__.py`](https://github.com/TomAugspurger/pandas/blob/0fdbfd34add4fe9940cd1991122e851d62824724/pandas/tests/extension/base/__init__.py) !

 - There are no PEP8 issues in the file [`pandas/tests/extension/base/interface.py`](https://github.com/TomAugspurger/pandas/blob/0fdbfd34add4fe9940cd1991122e851d62824724/pandas/tests/extension/base/interface.py) !

 - There are no PEP8 issues in the file [`pandas/tests/extension/base/printing.py`](https://github.com/TomAugspurger/pandas/blob/0fdbfd34add4fe9940cd1991122e851d62824724/pandas/tests/extension/base/printing.py) !

 - There are no PEP8 issues in the file [`pandas/tests/extension/decimal/array.py`](https://github.com/TomAugspurger/pandas/blob/0fdbfd34add4fe9940cd1991122e851d62824724/pandas/tests/extension/decimal/array.py) !

 - There are no PEP8 issues in the file [`pandas/tests/extension/decimal/test_decimal.py`](https://github.com/TomAugspurger/pandas/blob/0fdbfd34add4fe9940cd1991122e851d62824724/pandas/tests/extension/decimal/test_decimal.py) !

 - There are no PEP8 issues in the file [`pandas/tests/extension/json/array.py`](https://github.com/TomAugspurger/pandas/blob/0fdbfd34add4fe9940cd1991122e851d62824724/pandas/tests/extension/json/array.py) !

 - There are no PEP8 issues in the file [`pandas/tests/extension/json/test_json.py`](https://github.com/TomAugspurger/pandas/blob/0fdbfd34add4fe9940cd1991122e851d62824724/pandas/tests/extension/json/test_json.py) !

 - There are no PEP8 issues in the file [`pandas/tests/extension/test_integer.py`](https://github.com/TomAugspurger/pandas/blob/0fdbfd34add4fe9940cd1991122e851d62824724/pandas/tests/extension/test_integer.py) !

 - There are no PEP8 issues in the file [`pandas/tests/extension/test_interval.py`](https://github.com/TomAugspurger/pandas/blob/0fdbfd34add4fe9940cd1991122e851d62824724/pandas/tests/extension/test_interval.py) !

 - There are no PEP8 issues in the file [`pandas/tests/extension/test_period.py`](https://github.com/TomAugspurger/pandas/blob/0fdbfd34add4fe9940cd1991122e851d62824724/pandas/tests/extension/test_period.py) !

 - There are no PEP8 issues in the file [`pandas/tests/extension/test_sparse.py`](https://github.com/TomAugspurger/pandas/blob/0fdbfd34add4fe9940cd1991122e851d62824724/pandas/tests/extension/test_sparse.py) !

TomAugspurger commented:
```python
In [3]: integer_array([1, 2, 3])
Out[3]:
<IntegerArray>
[1, 2, 3]
Length: 3, dtype: Int64

In [4]: period_array(['2000', '2001'], freq='D')
Out[4]:
<PeriodArray>
[2000-01-01, 2001-01-01]
Length: 2, dtype: period[D]

In [5]: IntervalArray.from_breaks([1, 2, 3])
Out[5]:
<IntervalArray>
[(1, 2], (2, 3]]
Length: 2, dtype: interval[int64]

In [6]: integer_array([1, 2, 3] * 1000)
Out[6]:
<IntegerArray>
[1, 2, 3, 1, 2, 3, 1, 2, 3, 1,
 ...
 3, 1, 2, 3, 1, 2, 3, 1, 2, 3]
Length: 3000, dtype: Int64

```

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -653,15 +660,46 @@ def copy(self, deep=False):
+                               dtype=self.dtype)
+
+    @property
+    def _formatter(self):
```
Wouldn't it be possible to combine this with `_formatting_values` ?  

Or that makes it hard to use the  `format_object_summary` ?

TomAugspurger replied:
I gave up on sharing these. I think the order of operations is swapped between `format_object_summary` and `_formatting_values` (via internals).

IIUC, `format_object_summary` takes a whole array, truncates, then formats (need to get the length). But internals will truncate *before* calling `formatting_values`.

jorisvandenbossche replied:
Hmm, I understand that that's how our internals currently work, it's still a bit unfortunate to clutter our EA interface with it though ...

Would a not too big hack be a check inside `format_object_summary` for ExtensionArray, and in that case swap `obj` with `obj[selection]._formatting_values()` (where `selection` can just be the maximum values needed to select)

jreback replied:
@jorisvandenbossche 

> Hmm, I understand that that's how our internals currently work, it's still a bit unfortunate to clutter our EA interface with it though ...

its not unfortunate, rather its fortunate that we have a well test interface already. The EA interface IS lacking by-definition if you have to repeat code that we already have.


> would a not too big hack be a check inside format_object_summary for ExtensionArray, and in that case swap obj with obj[selection]._formatting_values() (where selection can just be the maximum values needed to select)

really try to avoid 'hacks'

TomAugspurger replied:
I think you misunderstood what Joris was saying. His point was it's unfortunate to have both `_formatting_values` and this new repr.

jorisvandenbossche replied:
> its not unfortunate, rather its fortunate that we have a well test interface already. The EA interface IS lacking by-definition if you have to repeat code that we already have.

Sorry Jeff, I don't understand your comment.

jorisvandenbossche replied:
> His point was it's unfortunate to have both _formatting_values and this new repr.

Small clarification: having both `_formatting_values` and `_formatter`. Having a default `__repr__` is great.

jorisvandenbossche replied:
And the reason that we in general should always be conservative in adding new methods here, is that this is all public interface (even if is a private attribute, it's part of the public EA interface)

jreback replied:
my comment means. that it is unfortunate that this code was repeated in the first place. The entire point of EA is to use our existing code that we *already* have as much as possible. I have complained about this before and will keep focusing on this here. Avoiding repeating the many may mistakes made in the repr here are paramount. Having repeated code is a distaster.

TomAugspurger replied:
@jorisvandenbossche how about we remove `_formatting_values` from the public interface, in favor of `_formatter`? Then in `ExtensionBlock._formatting_values`, we can do 

```python
def formatting_values(self):
    formatter = self.values._formatter
    return np.array([formatter(x) for x in self.values], dtype=object)
```

jorisvandenbossche replied:
Just to be sure (can't check right now): `_formatting_values` is only used for printing? And not for eg conversion to string in general (as that could give a performance penalty if the EA has a more efficient way to convert to strings)

And in the formatting code, are there some options that could influence the formatting (just thinking out loud, eg the datetime format)

TomAugspurger replied:
`_formatting_values` is only use is printing.

The trouble I'm running into now is an inconsistency between printing in Series and Index. Using (e.g.) datetimes, on master we quote values in the index, but not in the series.

```python
In [26]: pd.Series(pd.DatetimeIndex(['2001', '2002']))
Out[26]:
0   2001-01-01
1   2002-01-01
dtype: datetime64[ns]

In [27]: pd.DatetimeIndex(['2001', '2002'])
Out[27]: DatetimeIndex(['2001-01-01', '2002-01-01'], dtype='datetime64[ns]', freq=None)

In [28]: pd.Series(arr)
Out[28]:
0    '2000-01-01'
1    '2001-01-01'
dtype: period[D]

In [29]: pd.Index(arr)
Out[29]: PeriodIndex(['2000-01-01', '2001-01-01'], dtype='period[D]', freq='D')
```

Do we want to keep it like that? If so, `_formatter` would be a function that takes the "context" we're printing in?

jorisvandenbossche replied:
> Do we want to keep it like that? If so, _formatter would be a function that takes the "context" we're printing in (by "context", I mean whether we're inside a series / dataframe)?

I think I would personally be fine with dropping the quotes on the array repr as well, but for the Index repr it would of course be a bigger change ...

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -653,15 +660,46 @@ def copy(self, deep=False):
-    # Block-related methods
+    # Printing
     # ------------------------------------------------------------------------
+    def __repr__(self):
```
blank line above

jorisvandenbossche reviewed (commented):
Do you want to add one for DatetimeArray and TimedeltaArray here as well?

Do we need a mechanism to indicate which attributes to print in addition to length and dtype? (in case we want to keep printing the `freq` for DatetimeArray/Timedelta)Array

Is there any control over the number of elements shown?

TomAugspurger commented:
For DatetimeArray & Timedelta I was waiting to see what happens on https://github.com/pandas-dev/pandas/pull/23587. If @jbrockmendel reverts the repr changes before merging then I'll add it here. Otherwise I'll just delete them after that's merged.

>  (in case we want to keep printing the freq for DatetimeArray/Timedelta)Array

> Do we need a mechanism to indicate which attributes to print in addition to length and dtype? (in case we want to keep printing the freq for DatetimeArray/Timedelta)Array

Sure. I suppose that info doesn't belong in the dtype (right?) so we can add hooks for extra attrs.

> Is there any control over the number of elements shown?

At the array level, no. But I think following the option at `pd.options.display.max_seq_items` is the right thing to do. I'll add a note to the docs.

jorisvandenbossche commented:
The main difference for PeriodArray seems to be that the values are no longer quoted? (not a strong opinion here though, for Index reprs we use quotes and also numpy quotes dates)

TomAugspurger commented:
```diff
diff --git a/pandas/core/arrays/period.py b/pandas/core/arrays/period.py
index f6996f8e6..4bc7841fe 100644
--- a/pandas/core/arrays/period.py
+++ b/pandas/core/arrays/period.py
@@ -330,6 +330,10 @@ class PeriodArray(dtl.DatetimeLikeArrayMixin, ExtensionArray):
     def end_time(self):
         return self.to_timestamp(how='end')
 
+    @property
+    def _formatter(self):
+        return "'{}'".format
+
     def __setitem__(
             self,
             key,   # type: Union[int, Sequence[int], Sequence[bool]]
```

will quote. Gonna mixup the py2 issues.

TomAugspurger committed:
Deprecate formatting_values

TomAugspurger commented:
Update:

1. deprecated `ExtensionArray._formatting_values`
2. Change `_formatter` to a function that takes a flag for whether or we're printing inside a Series / DataFrame


```python
In [5]: integer_array([1, 2, None] * 1000)
Out[5]:
<IntegerArray>
[  1,   2, nan,   1,   2, nan,   1,   2, nan,   1,
 ...
 nan,   1,   2, nan,   1,   2, nan,   1,   2, nan]
Length: 3000, dtype: Int64

In [6]: IntervalArray.from_breaks(list(range(1000)))
Out[6]:
<IntervalArray>
[    (0, 1],     (1, 2],     (2, 3],     (3, 4],     (4, 5],     (5, 6],
     (6, 7],     (7, 8],     (8, 9],    (9, 10],
 ...
 (989, 990], (990, 991], (991, 992], (992, 993], (993, 994], (994, 995],
 (995, 996], (996, 997], (997, 998], (998, 999]]
Length: 999, dtype: interval[int64]

In [7]: period_array(['2000', '2001'] * 1000, freq='D')
Out[7]:
<PeriodArray>
['2000-01-01', '2001-01-01', '2000-01-01', '2001-01-01', '2000-01-01',
 '2001-01-01', '2000-01-01', '2001-01-01', '2000-01-01', '2001-01-01',
 ...
 '2000-01-01', '2001-01-01', '2000-01-01', '2001-01-01', '2000-01-01',
 '2001-01-01', '2000-01-01', '2001-01-01', '2000-01-01', '2001-01-01']
Length: 2000, dtype: period[D]
```

TomAugspurger committed:
test for warning

TomAugspurger committed:
compat

TomAugspurger commented:
It's part of the interface and was around since 0.23.

TomAugspurger committed:
na formatter

TomAugspurger committed:
clean

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into ea-repr

TomAugspurger committed:
wip

TomAugspurger committed:
more cleanup

TomAugspurger commented:
This are into a bit larger of a refactor... I removed {Categorica,Period,Interval}lArrayFormatter in favor of a generic ExtensionArrayFormatter.

EAs will get control over formatting of individual values by overriding `ExtensionArray._formatter`.

TomAugspurger committed:
update docs, type

TomAugspurger committed:
format

TomAugspurger committed:
try this

TomAugspurger commented on pandas/tests/frame/test_repr_info.py:
```diff
@@ -514,12 +514,12 @@ def test_repr_categorical_dates_periods(self):
-2 2011-01-01 11:00:00-05:00 2011-03
-3 2011-01-01 12:00:00-05:00 2011-04
-4 2011-01-01 13:00:00-05:00 2011-05"""
+        exp = """                         dt        p
```
I'm getting an extra space here, but don't know why yet. Will try to revert this if it's a bug.

jreback replied:
yeah this looks odd. no idea why.

TomAugspurger commented on pandas/io/formats/format.py:
```diff
@@ -1126,39 +1122,22 @@ def _format_strings(self):
-            # periods may contains different freq
-            values = Index(self.values, dtype='object').to_native_types()
-
-        formatter = self.formatter or (lambda x: '{x}'.format(x=x))
```
I *think* this is the cause of the extra space. Periods were incorrectly missing a space earlier.

On 0.23.4, notice that the period values start above the `e` in `dtype`, rather than above the `:` like the rest:

```python
In [9]: pd.Series(range(3))
Out[9]:
0    0
1    1
2    2
dtype: int64

In [10]: pd.Series(pd.period_range('2000', periods=3))
Out[10]:
0   2000-01-01
1   2000-01-02
2   2000-01-03
dtype: object

```

On this PR, periods start above the colon:

```python
In [7]: pd.Series(range(3))
Out[7]:
0    0
1    1
2    2
dtype: int64

In [8]: pd.Series(pd.period_range('2000', periods=3))
Out[8]:
0    2000-01-01
1    2000-01-02
2    2000-01-03
dtype: period[D]
```

TomAugspurger committed:
updates

TomAugspurger committed:
fixup interval

TomAugspurger committed:
py2 compat

jreback commented on pandas/core/arrays/base.py:
```diff
@@ -653,15 +660,60 @@ def copy(self, deep=False):
+        template = (
+            '<{class_name}>\n'
+            '{data}\n'
+            'Length: {length}, dtype: {dtype}'
```
do we need to define the “__unicode__”
we do this in Base for all pandas objects

jreback replied:
you are writing new code here but this should be consistent as well (it’s ok to change that too)
but to have a completely different impl is odd

TomAugspurger replied:
Fixed... I left the implementation in repr, and then encoded / decoded as needed in `__unicdoe__` and `__bytes__` if that's OK.

jreback commented on pandas/io/formats/printing.py:
```diff
@@ -381,7 +397,11 @@ def best_len(values):
-        summary += '],'
+
+        # right now close is either '' or ', '
+        # Now we want to include the ']', but not the maybe space.
```
why is this needed?

TomAugspurger replied:
Without it, you'd get something like

```python
In [2]: pd.core.arrays.period_array(['2000', '2001'], freq='D')
Out[2]:
<PeriodArray>
['2000-01-01', '2001-01-01'], 
Length: 2, dtype: period[D]
```

(notice the trailing coma after the ending `]` in the data section.


That's what we want for index classes:

```python
In [3]: pd.Index([1, 2, 3])
Out[3]: Int64Index([1, 2, 3], dtype='int64')
```

but not for EAs since we don't know if they're valid code:

jreback replied:
i think this is a very strange repr 
where is this based off of? why not use parens?
it’s also multi line 

this is reinventing the wheel again compared to what we do for index; and somewhat arbitrary

repr is very important for consistency and this is off

i would avoid the special casing and have it look a whole lot more like what we have now

jreback replied:
also none of this looks tested (meaning the special casing and so on)

jreback replied:
the valid code argument does not work here either

we already have not valid code in repr eg MultiIndex and IntervalIndex

it’s very hard to guarantee this 

but using parens and no angle brackets and commas between items would be a major improvement

TomAugspurger replied:
> also none of this looks tested (meaning the special casing and so on)
 
What special casing? I have 100% coverage for the diff when running it on the base repr tests.

TomAugspurger replied:
Why's that? Don't we want to reuse the common formatting code?

jreback replied:
so another difference this is highliting is that EA have the attributes on another line, while the Index does not (as they are args).

jreback commented on pandas/io/formats/printing.py:
```diff
@@ -283,9 +284,14 @@ def format_object_summary(obj, formatter, is_justify=True, name=None):
+    name : name, optional
         defaults to the class name of the obj
 
+        Pass ``False`` to indicate that subsequent lines should
```
what calls this with False? IOW what has a name that u don’t want to print

jreback replied:
can this be another parameter then? it seems like it is used for 2 purposes

TomAugspurger replied:
Done. Calling it `indent_for_name`.

jreback reviewed (changes requested):
looks really good @TomAugspurger thanks for doing this

TomAugspurger committed:
revert interval

TomAugspurger committed:
unicode, bytes

TomAugspurger commented:
FYI, I reverted the changes to the IntervalArray repr. Some doctests shared with IntervalIndex were failing.

TomAugspurger committed:
isort

TomAugspurger commented:
This is also a base repr for 3rd parties. Having reprs that look like code but isn't is a bad practice. We shouldn't put that on other libraries by default.

We are of course able to override the reps of our first party EAs if we really want to, but a foolish consistency and all that.

TomAugspurger committed:
py3 fixup

TomAugspurger committed:
fixup

TomAugspurger committed:
unicode

TomAugspurger committed:
unicode

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into ea-repr

TomAugspurger committed:
unicode

TomAugspurger committed:
lint

jreback commented on pandas/core/arrays/base.py:
```diff
@@ -653,15 +661,59 @@ def copy(self, deep=False):
+        from pandas.io.formats.printing import format_object_summary
+
+        template = (
+            u'<{class_name}>\n'
```
this is an abitrary repr here. This needs to justify why not using parens and not comma separated. -1 on having this different from the Index reprs that we already have.

TomAugspurger replied:
> this is an abitrary repr here. This needs to justify why not using parens and not comma separated.

Agreed. It's an arbitrary repr, and the justification is that we can't assume that it's valid code for the class inheriting it.

jorisvandenbossche replied:
@jreback see also the discussion in https://github.com/pandas-dev/pandas/issues/22846

jreback replied:
rather than creating a new arbitrary rep, let's just use what we have.

TomAugspurger commented on pandas/core/internals/blocks.py:
```diff
@@ -1951,7 +1951,19 @@ def _slice(self, slicer):
 
     def formatting_values(self):
-        return self.values._formatting_values()
+        # Deprecating the ability to override _formatting_values.
```
Fortunately, it's the only place we call it, so we can do this. It's added to the list.

TomAugspurger commented on pandas/tests/arrays/test_integer.py:
```diff
@@ -57,24 +57,28 @@ def test_dtypes(dtype):
 
 
-class TestInterface(object):
+def test_repr_array(data):
```
The base extension tests can only assert basic things like whether or not `...` shows up, whether the class name is present.

I'll update this to have the exact repr.

TomAugspurger committed:
update repr tests

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into ea-repr

TomAugspurger committed:
remove periodarray

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -49,6 +50,13 @@ class ExtensionArray(object):
 
     * _formatting_values
```
Remove (or mention it is deprecated) this?

jorisvandenbossche commented on pandas/core/internals/blocks.py:
```diff
@@ -1951,7 +1951,19 @@ def _slice(self, slicer):
+                "'ExtensionArray._formatting_values' is deprecated. "
+                "Specify 'ExtensionArray._formatter' instead."
+            )
+            warnings.warn(msg, FutureWarning, stacklevel=10)
```
DeprecationWarning instead of FutureWarning? (As this is not relevant for end user, only for developer)

TomAugspurger replied:
Good call.

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -653,15 +661,59 @@ def copy(self, deep=False):
+    # Printing
     # ------------------------------------------------------------------------
+    def __unicode__(self):
+        from pandas.io.formats.printing import format_object_summary
```
Can you add a comment here explaining that is basically defining `__repr__` but with some py2/3 compat things handled for you? Because EA authors might wonder why this is not `__repr__`

TomAugspurger replied:
I didn’t really realize the effects of this on downstream users of EA. This really should be implemented in repr, so that we don’t interact poorly with a custom str method.

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -653,15 +661,59 @@ def copy(self, deep=False):
+            The formatter this array is being rendered with. The formatter
+            may have a `.formatter` method already defined. By default, this
+            will be used if a `formatter` is passed, otherwise the formatter
+            is ``str``.
```
I didn't yet look in detail to the further changes in this PR, but from an EA author view, I find the above not very clear.

In what cases would a `formatter` be passed? As an external EA authord, do I need to care or can I ignore this?

TomAugspurger replied:
Will try to clarify. I believe formatter is passed when we’re boxed in a DataFrame, Series, or Index.

But for a top-level repr(array) no formatter is passed.

jorisvandenbossche commented:
General other question: should the default `_formatter` be `str` or `repr` ?

Eg, numpy uses `repr`:

```
In [31]: np.array([datetime.datetime(2012,1,1)])
Out[31]: array([datetime.datetime(2012, 1, 1, 0, 0)], dtype=object)
```

(which in the specific above case is useful to be able to see it are datetime.datetime objects)

I think for our own arrays, we typically use `str` (eg for Timestamp / Period, we don't include the class name in each scalar). 

Of course this is always easy to override by an EA author if the default is not what they want.

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into ea-repr

TomAugspurger committed:
FutureWarning -> DeprecationWarning

TomAugspurger commented:
I suppose it make sense for `__repr__` to use `repr` of the scalar.

TomAugspurger committed:
wip

TomAugspurger committed:
use repr

TomAugspurger committed:
fixup! use repr

TomAugspurger committed:
fixup! fixup! use repr

TomAugspurger commented:
Changes from earlier:

1. Changed FutureWarning -> DeprecationWarning
2. Changed the default `ExtensionArray._formatter` to return `repr`, not `str`.
3. Changed the implementation to define `__repr__` instead of `__unicde__`. (I didn't realize that `__unicode__` wasn't part of the Python data model).

jreback commented:
btw do you need the ``__bytes__`` ?

TomAugspurger commented:
Hmm I was thinking people may have expected it to follow the behavior of Categorical, but now I'm thinking that's not a good idea in general (for the default). I think that EAs, being array-like, should follow NumPy and offer something that can be dumped in to a memory view:

```python
In [12]: np.frombuffer(memoryview(bytes(np.array([1, 2]))), dtype=np.int64)
Out[12]: array([1, 2])
```

But the default can't know what actual data needs to be dumped, so we should (I think) just not implement it.

TomAugspurger committed:
remove bytes

TomAugspurger commented:
All green here. I think this is a good default repr for 3rd party EAs.

We can talk further about our own EAs, but

1. I value consistency with Categorical over consistency with Index reprs, since Categorical is an EA.
2. I'm not especially interested in changing the Categorical repr to match the index ones.

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into ea-repr

TomAugspurger commented:
Where do we stand here? Are we OK going forward with this as a base repr?

jorisvandenbossche commented:
I am fine with merging this.

@jreback can you respond to the comments in https://github.com/pandas-dev/pandas/issues/22846

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -44,10 +44,12 @@ class ExtensionArray(object):
 
-    * _formatting_values
+    * _formatter
+    * __repr__
```
Maybe also mention that `_formatter` is used to determine how the values are shown in Series / DataFrame as well? (I would even say that it is its main purpose)

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -653,15 +655,73 @@ def copy(self, deep=False):
+        # type: (Optional[ExtensionArrayFormatter]) -> Callable[[Any], str]
+        """Formatting function for scalar values.
+
+        This is used in the default '__repr__'. The formatting function
```
Something like "This is used when formatting the values in the repr of Series and DataFrame, and used in the default `__repr__` of the ExtensionArray" ?

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -653,15 +655,73 @@ def copy(self, deep=False):
+            is being rendered inside an Index, Series, or DataFrame, a
+            formatter will be provided. So if you want your objects to
+            render differently inside a Series from on its own, checking
+            with ``formatter is None`` is an option.
```
Are you only expected to simply do such check, to return something differently, or would you actually use some information of that formatter? 
(because for the latter, just a boolean flag would be sufficient?)

EDIT: ah, I see a couple of lines below that `formatter.formatter` is used.

TomAugspurger replied:
I'm actually not sure we can get in a situation where formatter will be specified to anything other than the default of None. There's a lot of indirection in this code right now :/

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -653,15 +655,73 @@ def copy(self, deep=False):
+            The default behavior depends on whether `formatter` is passed.
+
+            * When `formatter` is None, :func:`repr` is returned.
+            * When `formatter` is passed, ``formatter.formatter`` is used,
```
But what is ``formatter.formatter`` doing? 
(from the viewpoint of an EA author, I want to know what the default would be for my EA)

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -653,15 +655,73 @@ def copy(self, deep=False):
+            * When `formatter` is passed, ``formatter.formatter`` is used,
+              which falls back to :func:`repr` if that isn't specified.
+
+            In general, just returning :func:`repr` should be fine.
```
add ``"or :func:`str` "`` ? (I feel that also will be a common case, eg in geopandas I will want the string version, and not the repr)

jorisvandenbossche commented on pandas/core/arrays/period.py:
```diff
@@ -350,13 +350,10 @@ def to_timestamp(self, freq=None, how='start'):
+    def _formatter(self, formatter=None):
+        if formatter:
+            return formatter.formatter or str
+        return "'{}'".format
```
why not `str` like on the line above?

TomAugspurger replied:
This was to support quoting the periods in the array repr but not in the Series. If we want to quote in both places (or not) then we can have just a single one.

PeriodIndex does quote.

TomAugspurger replied:
https://github.com/pandas-dev/pandas/pull/23601#issuecomment-439882626

jreback replied:
is this the same as what DatetimeArray / Timedeltarray will do?

jreback commented:
I responded. -1  on the specific changes as outlined. The general approach is nice to actually re-use existing code.

TomAugspurger commented:
How should we proceed then?

On Thu, Nov 15, 2018 at 7:16 AM Jeff Reback <notifications@github.com>
wrote:

> I responded. -1 on the specific changes as outlined. The general approach
> is nice to actually re-use existing code.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/23601#issuecomment-439036371>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABQHIrhq6BDlXwCgtRnDKMM5Ok8Pv5oXks5uvWksgaJpZM4YWyCT>
> .
>

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into ea-repr

TomAugspurger committed:
simplify formatter

TomAugspurger commented:
Latest push simplifies `ExtensionArray._formatter` for the EA author. Now they just receive a keyword `boxed=False` for whether pandas is calling the `repr` to be printed inside a Series, DataFrame (or Index eventually).

We use that in PeriodArray to quote, or not, depending on the context.

TomAugspurger commented on pandas/core/arrays/base.py:
```diff
@@ -653,15 +655,65 @@ def copy(self, deep=False):
+
+        Parameters
+        ----------
+        boxed: bool, default False
```
We needed it to get the period tests passing. PeriodIndex quotes scalar values, but `Series[Period]` doesn't (which is fine, since the Series repr isn't supposed to be code).

jreback replied:
well, maybe fix that. This is completely inconsistent with the rest of the reprs.

TomAugspurger replied:
Datetime is the same.

```python
In [2]: pd.date_range('2017', periods=4)
Out[2]: DatetimeIndex(['2017-01-01', '2017-01-02', '2017-01-03', '2017-01-04'], dtype='datetime64[ns]', freq='D')

In [3]: pd.Series(pd.date_range('2017', periods=4))
Out[3]:
0   2017-01-01
1   2017-01-02
2   2017-01-03
3   2017-01-04
dtype: datetime64[ns]
```

jreback replied:
this is a Series repr, that is different from Index. Array is *much* more like Index.

TomAugspurger replied:
Responded in 
https://github.com/pandas-dev/pandas/pull/23601#issuecomment-439882626

TomAugspurger commented on pandas/core/arrays/base.py:
```diff
@@ -653,15 +655,65 @@ def copy(self, deep=False):
+        template = (
+            u'{class_name}'
+            u'{data}\n'
+            u'Length: {length}, dtype: {dtype}'
```
Why lower? It's uppercase in the Series repr.

jreback replied:
its lowercase everywhere (else)

TomAugspurger replied:
Everywhere else being index? Categorical uses a capital. I'd like to eventually use pieces of this for the categorical repr, and would rather not break that repr, so I think a capital makes more sense here.

jorisvandenbossche replied:
In index it is as part of the "keywords" in the constructor-resembling repr. There it indeed makes sense to have it lowercase. But here, I think capitalized is much more logical (it's the first item on that line), and consistent with Series and Categorical.

jreback replied:
ok, I guess if we match the EA's to Series, except for the quoting (e.g. quote in EA, but no change in Series, meaning no quoting), then Index is separate. I am still concerned with these slight differences however. 

E.g. even here, dtype is lowercase and Length is uppercase

TomAugspurger replied:
Keep in mind that any discussion of quoting whether or not to quote is up to the individual array.

Are you specifically talking about quoting within PeriodArray here? Do we plan to quote within DatetimeArray?

TomAugspurger commented on pandas/tests/extension/test_integer.py:
```diff
@@ -210,3 +210,7 @@ class TestNumericReduce(base.BaseNumericReduceTests):
 class TestBooleanReduce(base.BaseBooleanReduceTests):
     pass
+
+
```
What's the question? The'res a `TestPrinting` right below this line, which is the interface repr tests.

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into ea-repr

TomAugspurger committed:
Updates: misc

* whatsnew
* docstrings

TomAugspurger committed:
BUG: Fixed SparseArray formatter

We want to fall back to the implementation in formats.

TomAugspurger commented:
In terms of the repr I think we closer to Series, since neither is executable, the values needn’t be quoted. And we don’t want to quote here, since then you’d have ambiguity with a series of strings.


________________________________
From: Jeff Reback <notifications@github.com>
Sent: Monday, November 19, 2018 06:47
To: pandas-dev/pandas
Cc: Tom Augspurger; Mention
Subject: Re: [pandas-dev/pandas] Add default repr for EAs (#23601)


@jreback commented on this pull request.

________________________________

In pandas/core/arrays/base.py<https://github.com/pandas-dev/pandas/pull/23601#discussion_r234605325>:

> +                                     trailing_comma=False).rstrip()
+        class_name = u'<{}>\n'.format(self.__class__.__name__)
+        return template.format(class_name=class_name, data=data,
+                               length=len(self),
+                               dtype=self.dtype)
+
+    def _formatter(self, boxed=False):
+        # type: (bool) -> Callable[[Any], str]
+        """Formatting function for scalar values.
+
+        This is used in the default '__repr__'. The returned formatting
+        function receives instances of your scalar type.
+
+        Parameters
+        ----------
+        boxed: bool, default False


this is a Series repr, that is different from Index. Array is much more like Index.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/pandas-dev/pandas/pull/23601#discussion_r234605325>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ABQHIkERM8ytJkDLZa8rjvoXHP-XKAjzks5uwqhrgaJpZM4YWyCT>.

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into ea-repr

TomAugspurger committed:
correct boxing

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into ea-repr

TomAugspurger commented:
I think the outstanding discussions are

1. capital vs. lowercase "L" in "Length": 
https://github.com/pandas-dev/pandas/pull/23601#discussion_r234464865
2. The `boxed` argument passed to the `_formatter`: 
https://github.com/pandas-dev/pandas/pull/23601#discussion_r234464938 (reply at 
https://github.com/pandas-dev/pandas/pull/23601#issuecomment-439882626)

jorisvandenbossche commented:
Question related to the boxed argument / quotes or not: assume we have a StringArray, how would we want it to look like? Quoted or not quoted? 
In a Series, we don't quote strings, but in such an array repr, I would actually be fine with having it quoted? So in a similar sense, it is maybe fine to have the Periods quoted in the Array repr?

TomAugspurger commented:
> Question related to the boxed argument / quotes or not: assume we have a StringArray, how would we want it to look like? Quoted or not quoted?

I would say quoted. I don't have a strong preference for PeriodArray. In either case, it shouldn't be ambiguous what the contents are, since they're type-specific containers. I opted for no quotes since they seemed unnecessary.

jreback reviewed (changes requested):
commented above

TomAugspurger commented:
Do people have preferences on quoting or not in Period/Datetime/TimedeltaArray?

jreback commented:
I think we *should* quote in the arrays (e.g. like we do in Index), but not in Series.

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into ea-repr

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into ea-repr

TomAugspurger commented:
I haven't pushed the short-repo changes yet, in case we want to defer on that.

```
In [4]: pd.core.arrays.period_array(['2000', '2001', None], freq='D')
Out[4]:
<PeriodArray>
['2000-01-01', '2001-01-01', 'NaT']
Length: 3, dtype: period[D]

In [5]: pd.core.arrays.period_array(['2000', '2001', None] * 100, freq='D')
Out[5]:
<PeriodArray>
['2000-01-01', '2001-01-01',        'NaT', '2000-01-01', '2001-01-01',
        'NaT', '2000-01-01', '2001-01-01',        'NaT', '2000-01-01',
 ...
        'NaT', '2000-01-01', '2001-01-01',        'NaT', '2000-01-01',
 '2001-01-01',        'NaT', '2000-01-01', '2001-01-01',        'NaT']
Length: 300, dtype: period[D]

In [6]: pd.core.arrays.integer_array([1, 2, None])
Out[6]:
<IntegerArray>
[1, 2, NaN]
Length: 3, dtype: Int64

In [7]: pd.core.arrays.integer_array([1, 2, None] * 1000)
Out[7]:
<IntegerArray>
[  1,   2, NaN,   1,   2, NaN,   1,   2, NaN,   1,
 ...
 NaN,   1,   2, NaN,   1,   2, NaN,   1,   2, NaN]
Length: 3000, dtype: Int64
```

jreback commented:
https://github.com/pandas-dev/pandas/pull/23601#issuecomment-442626358 would be ok with me (I still don't love the angle brackets) but this makes consistent EA reprs which is good.

jorisvandenbossche commented:
Given your comment @jreback , I think this is ready to be merged?

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -655,15 +657,65 @@ def copy(self, deep=False):
+        ----------
+        boxed: bool, default False
+            An indicated for whether or not your array is being printed
+            within a Series, DataFrame, or Index (True), or just by
```
Doesn't Index also use `False` like Array? (as I thought this was to satisfy the difference between Series/DataFrame vs Index)

TomAugspurger replied:
Index is hypothetical at this point, until we support EA-backed indexes. This isn't currently called by our index reprs. So we can choose to define it how we want.  Or we can make this a `box=None/Series/Index` argument.

jorisvandenbossche replied:
For external EAs, yes. But eg PeriodArray is stored in an Index, and there `box` is used for this purpose, no?

TomAugspurger replied:
It was not. Done in 3825aeb7f61b4d350888882689bce521b4a603ce

jreback replied:
I would add this arg to the Index formatters as well for compatiblity.

TomAugspurger replied:
I'm not sure I follow the need for adding it to the index formatters, since indexes are boxed by definition.

jorisvandenbossche commented:
@TomAugspurger I was thinking for a moment again on the quoting issue.

It would be nice to have a somewhat general rule about this, and also to reflect this in the base repr.

From observing the results, it seems that - mostly - we use a `str` representation in Series / DataFrame, and an (abbreviated) `repr` in Index / Array. With 'abbreviated' I mean for eg Timestamp, we use `'2012-01-01'` instead of `Timestamp('2012-01-01')` (so leaving out the object type) since the dtype already makes it clear they are all of that object type.

If we like this general pattern, we could also do this for the default repr (and sorry, that goes against what I commented earlier https://github.com/pandas-dev/pandas/pull/23601#issuecomment-437891679 and what you changed). So instead of 

```
def _formatter(self, boxed=False):
    return repr
```

we could have

```
def _formatter(self, boxed=False):
    if boxed:
        return str
    return repr
```

That would also fit a possible StringArray to not quote it in Series/DataFrame, but to show it quoted in the array repr.

Anyway, since this is configurable, and I think none of the internal ones inherits the base `_formatter` implementation, it is not that important, and certainly does not need to hold up this PR.

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into ea-repr

TomAugspurger committed:
Use Array formatter in PeriodIndex

TomAugspurger committed:
Use repr / str

TomAugspurger commented:
Implemented 
https://github.com/pandas-dev/pandas/pull/23601#issuecomment-443429153 in 2a60c15e755bb64b582875c4548cd38b6a92cf70.

jreback commented on doc/source/whatsnew/v0.24.0.rst:
```diff
@@ -1114,6 +1115,7 @@ Deprecations
 - The methods :meth:`Series.str.partition` and :meth:`Series.str.rpartition` have deprecated the ``pat`` keyword in favor of ``sep`` (:issue:`22676`)
 - Deprecated the `nthreads` keyword of :func:`pandas.read_feather` in favor of
   `use_threads` to reflect the changes in pyarrow 0.11.0. (:issue:`23053`)
+- :meth:`ExtensionArray._formatting_values` is deprecated. Use `ExtensionArray._formatter` instead. (:issue:`23601`)
```
can you add the ``:meth:`` syntax for ``_formatter`` (and maybe remove for ``_formatting_values``. is this removed from ``api.rst`` (for _formatter_values) and added for ``_formatter``)?

TomAugspurger replied:
Fixing the `:meth:`. There's a problem with api.rst excluding "private" methods: https://github.com/pandas-dev/pandas/issues/24067

jreback commented on pandas/core/arrays/base.py:
```diff
@@ -655,15 +657,69 @@ def copy(self, deep=False):
+        # repr does. So we include a newline in our template, and strip
+        # any trailing newlines from format_object_summary
+        data = format_object_summary(self, self._formatter(), name=False,
+                                     trailing_comma=False).rstrip()
```
I really find this ``trailing_comma`` argument bothersome. maybe it makes sense to name this like 'multi-line' or something, its seems that we built the ``format_object_summary`` to be for Index (which is true), then extended this (also true), so maybe it needs minor adjustment *inside* the method, rather than an option like this.

TomAugspurger replied:
I've removed `trailing_comma` in favor of a `.strip(', \n')` after formatting in a3b1c92347b469740a7d7ad20f9cb97b357086dd. Don't really have a preference for which we do.

jreback commented on pandas/core/arrays/categorical.py:
```diff
@@ -494,6 +494,10 @@ def _constructor(self):
         return Categorical(scalars, dtype=dtype)
 
+    def _formatter(self, boxed=False):
+        # backwards compat with old printing.
```
not sure what this comment means.

jreback commented on pandas/core/arrays/sparse.py:
```diff
@@ -1735,6 +1735,9 @@ def __unicode__(self):
             index=printing.pprint_thing(self.sp_index))
 
+    def _formatter(self, boxed=False):
+        return None
```
what does this mean / do?

TomAugspurger replied:
Falls back to the default for whatever `*Formatter` is printing the array. So this will end up being something like `FloatFormatter._value_formattter()`, or whatever, depending on the dtype. I'll add a comment to this (and clarify the one for categorical)

TomAugspurger replied:
Hopefully clarified in a7ef10406be10e21605e32b05fe196711f46bed4

jreback replied:
yep

jreback commented on pandas/io/formats/printing.py:
```diff
@@ -283,9 +284,14 @@ def format_object_summary(obj, formatter, is_justify=True, name=None):
 
+        Pass ``False`` to indicate that subsequent lines should
+        not be indented to align with the name.
+    trailing_comma : bool, default True
```
see if we can have a more meaningful name here, e.g. this is a multi-line repr yes?

TomAugspurger replied:
Not quite... since we could have a one-liner with or without a trailing comma

```
In [6]: pidx
Out[6]: PeriodIndex(['2000-01-01'], dtype='period[D]', freq='D')
```

So that has a trailing comma after values, since it's inside the `PeriodIndex` constructor. But 


```python
In [8]: pidx.array
Out[8]:
<PeriodArray>
['2000-01-01']
Length: 1, dtype: period[D]
```

doesn't have one, since it's not inside the constructor. I don't know if the hypothetical EA-one-line repr needs one or not.

```
# <PeriodArray>['2000-01-01'] dtype: period[D]
# <PeriodArray>['2000-01-01'], dtype: period[D]
```

TomAugspurger replied:
If you prefer, I can remove `trailing_comma` entirely, and just `.rstrip(', \n')` in the caller that doesn't need it.

jreback replied:
yeah maybe that's better

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into ea-repr

TomAugspurger committed:
Update for review

* docs
* removed overloading of name=False
* added indent_for_name

TomAugspurger committed:
REF: removed trailing_comma argument

jreback commented:
can you rebase

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into ea-repr

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into ea-repr

jreback commented on pandas/core/indexes/period.py:
```diff
@@ -503,7 +503,7 @@ def __array_wrap__(self, result, context=None):
     @property
     def _formatter_func(self):
-        return lambda x: "'%s'" % x
+        return self.array._formatter(boxed=False)
```
why is this the only index sublcass that you need to do this for?

TomAugspurger replied:
It's the only extension-array backed index that's using the formatting right now.

- CategoricalIndex always uses the default formatter for the underlying categories (since it can be a container for any type, it dispatches the formatting).
- IntervalIndex /IntervalArray don't use this formatting

Datetime / TImedelta will use this too, so this will be pushed up into `DatetimeIndexOpsMixin` later.

jreback commented on pandas/tests/arrays/test_integer.py:
```diff
@@ -57,24 +57,27 @@ def test_dtypes(dtype):
+    data = integer_array([1, 2, None] * 1000)
+    expected = (
+        "<IntegerArray>\n"
+        "[  1,   2, NaN,   1,   2, NaN,   1,   2, NaN,   1,\n"
```
these are somehow justified?

TomAugspurger replied:
Justified, as in "vertically aligned"? Here's the repr

```python
In [10]: data
Out[10]:
<IntegerArray>
[  1,   2, NaN,   1,   2, NaN,   1,   2, NaN,   1,
 ...
 NaN,   1,   2, NaN,   1,   2, NaN,   1,   2, NaN]
Length: 3000, dtype: Int64
```

The NaN pattern there makes the formatting a bt strange, but I think unavoidable.

jreback replied:
no this is ok, was just wondering about

jreback reviewed (commented):
couple of comments

TomAugspurger commented on pandas/io/formats/format.py:
```diff
@@ -1121,39 +1117,22 @@ def _format_strings(self):
-    def __init__(self, values, *args, **kwargs):
-        GenericArrayFormatter.__init__(self, values, *args, **kwargs)
+        if is_categorical_dtype(values.dtype):
+            # Categorical is special for now, so that we can preserve tzinfo
```
That depends on whether we're willing to change `__array__` for datetime. I'm writing up an issue now to discuss that specific point.

TomAugspurger replied:
https://github.com/pandas-dev/pandas/issues/23569#issuecomment-444091967 for that.

TomAugspurger commented:
I'm not sure how hard it would be, but in

```python
<IntegerArray>
[  1,   2, NaN,   1,   2, NaN,   1,   2, NaN,   1,
 ...
 NaN,   1,   2, NaN,   1,   2, NaN,   1,   2, NaN]
Length: 3000, dtype: Int64
```

I don't like that we devote an entire line just  to `...` I wonder if we could instead do

```python
<IntegerArray>
[  1,   2, NaN,   1,   2, NaN,   1,   2, NaN, ...,
 ...,   1,   2, NaN,   1,   2, NaN,   1,   2, NaN]
Length: 3000, dtype: Int64
```

Although, that makes the continuation harder to see (for me)...

jreback commented:
this is fine: https://github.com/pandas-dev/pandas/pull/23601#issuecomment-444086330

we do this elsewhere, try this with a large Array and the own line makes a lot of sense.

TomAugspurger commented:
👍 all good then?

jreback commented:
thanks a lot @TomAugspurger ; this consolidates a lot of disparate code!

simonjayhawkins commented on pandas/io/formats/format.py:
```diff
@@ -1121,39 +1117,22 @@ def _format_strings(self):
 
-    def _format_strings(self):
-        fmt_values = format_array(self.values.get_values(), self.formatter,
+        fmt_values = format_array(array,
```
@TomAugspurger : i'm struggling to resolve some formatting issues. what is the reason for calling `format_array` here. As far as I can tell is looping back round to create a `GenericArrayFormatter` instance with a `formatter` specified to pick up the display options.

simonjayhawkins replied:
i guess, to be more succinct, why is `super()._format_strings()` not used?

jorisvandenbossche replied:
I am not that familiar with this code, but from a quick look: calling `super()._format_strings()` would be different, as this would call `GenericArrayFormatter._format_strings`, while the generic `format_array` can still result in using custom formatters like `Datetime64(TZ)Formatter` or `Timedelta64Formatter`, depending on what the values of the underlying EA are.

jorisvandenbossche replied:
Although that most of those custom Formatter classes don't do much special *if* formatter is specified. 

Eg `Datetime64Formatter` has this in `_format_strings`:

https://github.com/pandas-dev/pandas/blob/181f972d2f4e6bd5c8c33721175f68188cfa1a0d/pandas/io/formats/format.py#L1174-L1175

simonjayhawkins replied:
so if `ExtensionArrayFormatter` is not inheriting from `GenericArrayFormatter` but calling `format_array` to dispatch to another `...ArrayFormatter` class, why wouldn't the logic in `ExtensionArrayFormatter` be in `format_array`?

simonjayhawkins commented on pandas/core/arrays/integer.py:
```diff
@@ -268,6 +265,13 @@ def _from_sequence(cls, scalars, dtype=None, copy=False):
+    def _formatter(self, boxed=False):
+        def fmt(x):
+            if isna(x):
+                return 'NaN'
```
should `NaN` have been hardcoded here?

jorisvandenbossche replied:
Why not? This is used for the Integer data display, where we currently use NaN.

(whether we should use rather 'NA' instead of 'NaN', that's another question)

simonjayhawkins replied:
just wondering whether it would be a problem with `to_string(na_rep=...)`. will do some tests.

jorisvandenbossche replied:
Ah, yes, that's a good reason. But in general, this `_formatter` does not follow display options at all, is that correct? 
In which case this is something to think about in general.
